src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+        {
+            if (Math.Abs(preTotal) < Math.Abs(parent) * _anomalyDeltaThreshold)
+            {
+                return false;
+            }
+
+            return Math.Abs(pre) / Math.Abs(current) > _anomalyPreDeltaThreshold;
+        }
+
+        private bool ShouldSeperateAnomaly(double total, double parent, int totalSize, int size)
+        {
+            if (Math.Abs(total) < Math.Abs(parent) * _anomalyDeltaThreshold)
+            {
+                return false;
+            }
+
+            if (size == totalSize && size == 1)
+            {
+                return true;
+            }
+
+            return size <= totalSize * _anomalyRatioThreshold;
+        }
+
+        private double GetDimensionEntropy(Dictionary<string, int> pointDis, Dictionary<string, int> anomalyDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double entropy = 0;
+
+            foreach (string key in anomalyDis.Keys)
+            {
+                double dimEntropy = GetEntropy(pointDis[key], anomalyDis[key]);
+                entropy += dimEntropy * pointDis[key] / total;
+            }
+
+            return entropy;
+        }
+
+        private double GetDimensionInstrinsicValue(Dictionary<string, int> pointDis)
+        {
+            int total = GetTotalNumber(pointDis);
+            double instrinsicValue = 0;
+
+            foreach (string key in pointDis.Keys)
+            {
+                instrinsicValue -= Log2((double)pointDis[key] / total) * (double)pointDis[key] / total;
+            }
+
+            return instrinsicValue;
+        }
+
+        private int GetTotalNumber(Dictionary<string, int> distribution)
+        {
+            int total = 0;
+            foreach (int num in distribution.Values)
+            {
+                total += num;
+            }
+            return total;
+        }
+
+        private void UpdateDistribution(Dictionary<string, int> distribution, List<Point> points, string dimKey)
+        {
+            foreach (Point point in points)
+            {
+                string dimVal = (string)point.Dimension[dimKey];
+                if (!distribution.ContainsKey(dimVal))
+                {
+                    distribution.Add(dimVal, 0);
+                }
+                distribution[dimVal] = distribution[dimVal] + 1;
+            }
+        }
+
+        private double Log2(double val)
+        {
+            if (Double.IsNaN(val))
+            {
+                return 0;
+            }
+
+            return Math.Log(val) / Math.Log(2);
+        }
+
+        private static bool ContainsAll(Dictionary<string, Object> bigDic, Dictionary<string, Object> smallDic)
+        {
+            foreach (var item in smallDic)
+            {
+                if (!bigDic.ContainsKey(item.Key))
+                {
+                    return false;
+                }
+
+                if (bigDic.ContainsKey(item.Key) && !bigDic[item.Key].Equals(smallDic[item.Key]))
+                {
```

Comment:
You don't need the first ContainsKey check here becuase you are returning above. 
Also, please consider using LINQ techniques as suggested in other cases above.

Reply:
> You don't need the first ContainsKey check here becuase you are returning above.
> Also, please consider using LINQ techniques as suggested in other cases above.

I think the first ContainsKey check is needed as all the keys in small dictionary should be contained in big dictionary

Reply:
Sorry, I meant to say you don't need the second ContainsKey call. If item.Key is not present in the bigDictionary you return false. So when you get to line 677, you can be usre that item.Key is present in the bigDictionary and you don't need to check if bigDictionary.ContainsKey(item.Key)

---
In reply to: [422135511](https://github.com/dotnet/machinelearning/pull/4925#discussion_r422135511) [](ancestors = 422135511)

Reply:
> Sorry, I meant to say you don't need the second ContainsKey call. If item.Key is not present in the bigDictionary you return false. So when you get to line 677, you can be usre that item.Key is present in the bigDictionary and you don't need to check if bigDictionary.ContainsKey(item.Key)
> 
> In reply to: [422135511](https://github.com/dotnet/machinelearning/pull/4925#discussion_r422135511) [](ancestors = 422135511)

Totally agree with you, will update